File: library/src/main/java/com/github/library/oauth2/Access.java
package com.github.library.oauth2;


import java.io.IOException;

import androidx.collection.ArrayMap;

import com.github.library.Interface.AuthorizationParams;
import com.github.library.enums.AuthType;
import com.github.library.helper.general;
import com.github.library.model.OAuthResponse;

import okhttp3.Credentials;
import okhttp3.FormBody;

import okhttp3.Request;
import okhttp3.Response;

class Access implements AuthorizationParams {

    static OAuthResponse refreshAccessToken(String token, OAuth2Client oAuth2Client) throws IOException {
        FormBody.Builder formBodyBuilder = new FormBody.Builder().add(POST_REFRESH_TOKEN, token);
        general.postAddIfValid(formBodyBuilder, oAuth2Client.getFieldsAsMap());

        Request.Builder request = new Request.Builder()
                .url(oAuth2Client.getSite())
                .post(formBodyBuilder.build());

        addHeader(request, oAuth2Client);

        return getTokenFromResponse(oAuth2Client, request.build(), oAuth2Client.getAuthType());
    }

    static OAuthResponse getAccessToken(OAuth2Client oAuth2Client) throws IOException {
        FormBody.Builder formBodyBuilder = new FormBody.Builder();
        //TODO check it later
//        formBodyBuilder.addEncoded("grant_type","client_credentials");
        general.postAddIfValid(formBodyBuilder, oAuth2Client.getFieldsAsMap());
        Request.Builder request = new Request
                .Builder()
                .url(oAuth2Client.getSite())
                .post(formBodyBuilder.build());

        addHeader(request, oAuth2Client);
        return getTokenFromResponse(oAuth2Client, request.build(), oAuth2Client.getAuthType());
    }

    private static OAuthResponse getTokenFromResponse(OAuth2Client oAuth2Client, Request request, AuthType type) throws IOException {

        Response response = oAuth2Client.getOkHttpClient()
                .newBuilder()
                .build()
                .newCall(request.newBuilder()
                        .header(HEADER_AUTHORIZATION, Credentials.basic(oAuth2Client.getClientId(), oAuth2Client.getClientSecret()))
                        .build())

                .execute();

        return new OAuthResponse(response);
    }

    private static void addHeader(Request.Builder request, OAuth2Client oAuth2Client) {
        ArrayMap<String, String> header = oAuth2Client.getHeaders();
        for (int index = 0; index < header.size(); index++) {
            String key = header.keyAt(index);
            String value = header.get(key);
            assert value != null;
            request.addHeader(key, value);
        }
    }
}

File: library/src/main/java/com/github/library/oauth2/Access.kt
package com.github.library.oauth2


import com.github.library.helper.AuthorizationParams
import com.github.library.enums.AuthType
import com.github.library.helper.General
import com.github.library.model.OAuthResponse
import okhttp3.Credentials
import okhttp3.FormBody
import okhttp3.Request
import java.io.IOException

internal class Access  {

    companion object {
        @Throws(IOException::class)
        fun refreshAccessToken(token: String, oAuth2Client: OAuth2Client): OAuthResponse {
            val formBodyBuilder = FormBody.Builder().add(AuthorizationParams.POST_REFRESH_TOKEN, token)
            General.postAddIfValid(formBodyBuilder, oAuth2Client.fieldsAsMap)
            val request = Request.Builder()
                    .url(oAuth2Client.site)
                    .post(formBodyBuilder.build())
            addHeader(request, oAuth2Client)
            return getTokenFromResponse(oAuth2Client, request.build(), oAuth2Client.authType)
        }

        @Throws(IOException::class)
        fun getAccessToken(oAuth2Client: OAuth2Client): OAuthResponse {
            val formBodyBuilder = FormBody.Builder()
            General.postAddIfValid(formBodyBuilder, oAuth2Client.fieldsAsMap)
            val request = Request.Builder()
                    .url(oAuth2Client.site)
                    .post(formBodyBuilder.build())

            addHeader(request, oAuth2Client)
            return getTokenFromResponse(oAuth2Client, request.build(), oAuth2Client.authType)
        }

        @Throws(IOException::class)
        private fun getTokenFromResponse(oAuth2Client: OAuth2Client, request: Request, type: AuthType): OAuthResponse {
            val response = oAuth2Client.okHttpClient
                    .newBuilder()
                    .build()
                    .newCall(request.newBuilder()
                            .header(
                                    AuthorizationParams.HEADER_AUTHORIZATION,
                                    Credentials.basic(oAuth2Client.clientId, oAuth2Client.clientSecret)
                            )
                            .build())

                    .execute()
            return OAuthResponse(response)
        }

        private fun addHeader(request: Request.Builder, oAuth2Client: OAuth2Client) {
            val header = oAuth2Client.headers
            for (index in 0 until header.size) {
                val key = header.keyAt(index)
                val value = header[key]!!
                request.addHeader(key, value)
            }
            val grantedType = oAuth2Client.authModel.grantType
            request.addHeader("grant_type",grantedType)
        }
    }
}
